federation-js/src/composition/composeAndValidate.ts

```diff
@@ -7,13 +7,38 @@ import {
 import { ServiceDefinition } from './types';
 import { normalizeTypeDefs } from './normalize';
 import { printComposedSdl } from '../service/printComposedSdl';
+import { GraphQLError, GraphQLSchema } from 'graphql';
 
-export function composeAndValidate(serviceList: ServiceDefinition[]) {
+export type CompositionResult = CompositionFailure | CompositionSuccess;
+
+// Yes, it's a bit awkward that we still return a schema when errors occur.
+// This is old behavior that I'm choosing not to modify for now.
+export interface CompositionFailure {
+  /** @deprecated Use composedSdl instead */
+  schema: GraphQLSchema;
+  errors: GraphQLError[];
+}
+
+export interface CompositionSuccess {
+  /** @deprecated Use composedSdl instead */
+  schema: GraphQLSchema;
+  composedSdl: string;
+}
```

Comment:
Perhaps this is no longer the way it needs to be done and maybe it's a result of the type predicates you've put in place, but I'm surprised the `CompositionSuccess` type doesn't have to explicitly have `errors?: undefined` — and conversely — the `CompositionFailure` doesn't have to have `composedSdl?: undefined`.  (I've always found this odd behavior in TypeScript to make sure these aren't commingled, but I seem to recall it in a few places in our code.)

Reply:
I was shocked to find that I could return, for example:
`{ schema, errors: [], composedSdl: ''}`
...without a TS error.

I thought the disjointedness of the two objects would be enforced by TS, but it seems that's not the case.

Implemented your suggestion in 5170542 and that seems to do the trick. Also updated the predicate to handle the new potential existence of the key with an `undefined` value.